// @pm2/io/build/main/census/plugins/net.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.NetPlugin = void 0;
const core_1 = require("@opencensus/core");
const shimmer = require("shimmer");
class NetPlugin extends core_1.BasePlugin {
    constructor(moduleName) {
        super(moduleName);
    }
    applyPatch() {
        this.logger.debug('applying patch to %s@%s', this.moduleName, this.version);
        if (this.moduleExports && this.moduleExports.Server &&
            this.moduleExports.Server.prototype) {
            shimmer.wrap(this.moduleExports.Server.prototype, 'emit', this.getPatchIncomingRequestFunction());
        }
        else {
            this.logger.error('Could not apply patch to %s.emit. Interface is not as expected.', this.moduleName);
        }
        return this.moduleExports;
    }
    applyUnpatch() {
        if (this.moduleExports && this.moduleExports.Server &&
            this.moduleExports.Server.prototype) {
            shimmer.unwrap(this.moduleExports.Server.prototype, 'emit');
        }
        else {
            this.logger.error('Could not unapply patch to %s.emit. Interface is not as expected.', this.moduleName);
        }
    }
    getPatchIncomingRequestFunction() {
        return (original) => {
            const plugin = this;
            return function incomingRequest(event, ...args) {
                if (event !== 'connection') {
                    return original.apply(this, arguments);
                }
                const socket = args[0];
                plugin.logger.debug('%s plugin incomingRequest', plugin.moduleName);
                const traceOptions = {
                    name: 'socket',
                    kind: core_1.SpanKind.SERVER,
                    spanContext: undefined
                };
                return plugin.tracer.startRootSpan(traceOptions, rootSpan => {
                    if (!rootSpan)
                        return original.apply(this, arguments);
                    plugin.tracer.wrapEmitter(socket);
                    const address = socket.address();
                    if (typeof address === 'string') {
                        rootSpan.addAttribute('net.address', address);
                    }
                    else {
                        rootSpan.addAttribute('net.host', address.address);
                        rootSpan.addAttribute('net.port', address.port);
                        rootSpan.addAttribute('net.family', address.family);
                    }
                    socket.on('error', (err) => {
                        rootSpan.addAttribute('net.error', err.message);
                    });
                    const originalEnd = socket.end;
                    socket.end = function () {
                        if (rootSpan.ended === false) {
                            rootSpan.end();
                        }
                        return originalEnd.apply(this, arguments);
                    };
                    socket.on('close', () => {
                        if (rootSpan.ended === false) {
                            rootSpan.end();
                        }
                    });
                    return original.apply(this, arguments);
                });
            };
        };
    }
}
exports.NetPlugin = NetPlugin;
exports.plugin = new NetPlugin('net');
