// @pm2/io/build/main/census/plugins/vue.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.VuePlugin = void 0;
const core_1 = require("@opencensus/core");
const shimmer = require("shimmer");
var RendererType;
(function (RendererType) {
    RendererType["NORMAL"] = "normal";
    RendererType["BUNDLE"] = "bundle";
})(RendererType || (RendererType = {}));
class VuePlugin extends core_1.BasePlugin {
    constructor(moduleName) {
        super(moduleName);
        this.rendererResults = [];
    }
    applyPatch() {
        this.logger.debug('Patched vue-server-renderer');
        if (this.moduleExports) {
            this.logger.debug('patching vue-server-renderer.prototype.createRenderer');
            shimmer.wrap(this.moduleExports, 'createRenderer', this.getPatchCreateRenderer(RendererType.NORMAL));
            this.logger.debug('patching vue-server-renderer.prototype.createBundleRenderer');
            shimmer.wrap(this.moduleExports, 'createBundleRenderer', this.getPatchCreateRenderer(RendererType.BUNDLE));
        }
        return this.moduleExports;
    }
    applyUnpatch() {
        shimmer.unwrap(this.moduleExports, 'createRenderer');
        shimmer.unwrap(this.moduleExports, 'createBundleRenderer');
        for (let result of this.rendererResults) {
            shimmer.unwrap(result, 'renderToString');
        }
    }
    getPatchCreateRenderer(type) {
        const plugin = this;
        return function createRendererWrap(original) {
            return function create_renderer_trace() {
                const result = original.apply(this, arguments);
                plugin.logger.debug(`patching ${type}.renderToString`);
                shimmer.wrap(result, 'renderToString', plugin.getPatchRenderToString(type));
                plugin.rendererResults.push(result);
                return result;
            };
        };
    }
    getPatchRenderToString(type) {
        const plugin = this;
        return function renderToStringWrap(original) {
            return function render_string_trace() {
                if (!plugin.tracer.currentRootSpan) {
                    return original.apply(this, arguments);
                }
                const span = plugin.tracer.startChildSpan(`vue-renderer`, core_1.SpanKind.CLIENT);
                if (span === null)
                    return original.apply(this, arguments);
                const promise = original.apply(this, arguments);
                if (promise instanceof Promise) {
                    promise.then(plugin.patchEnd(span)).catch(plugin.patchEnd(span));
                }
                return promise;
            };
        };
    }
    patchEnd(span) {
        const plugin = this;
        const patchedEnd = function (err) {
            if (plugin.options.detailedCommands === true && err instanceof Error) {
                span.addAttribute('error', err.message);
            }
            if (span.ended === false) {
                span.end();
            }
        };
        return this.tracer.wrap(patchedEnd);
    }
}
exports.VuePlugin = VuePlugin;
const plugin = new VuePlugin('vue-server-renderer');
exports.plugin = plugin;
