// @pm2/io/build/main/census/exporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomCensusExporter = void 0;
const serviceManager_1 = require("../serviceManager");
const core_1 = require("@opencensus/core");
const default_config_1 = require("./config/default-config");
const constants_1 = require("./constants");
var CanonicalCodeString;
(function (CanonicalCodeString) {
    CanonicalCodeString["OK"] = "OK";
    CanonicalCodeString["CANCELLED"] = "CANCELLED";
    CanonicalCodeString["UNKNOWN"] = "UNKNOWN";
    CanonicalCodeString["INVALID_ARGUMENT"] = "INVALID_ARGUMENT";
    CanonicalCodeString["DEADLINE_EXCEEDED"] = "DEADLINE_EXCEEDED";
    CanonicalCodeString["NOT_FOUND"] = "NOT_FOUND";
    CanonicalCodeString["ALREADY_EXISTS"] = "ALREADY_EXISTS";
    CanonicalCodeString["PERMISSION_DENIED"] = "PERMISSION_DENIED";
    CanonicalCodeString["RESOURCE_EXHAUSTED"] = "RESOURCE_EXHAUSTED";
    CanonicalCodeString["FAILED_PRECONDITION"] = "FAILED_PRECONDITION";
    CanonicalCodeString["ABORTED"] = "ABORTED";
    CanonicalCodeString["OUT_OF_RANGE"] = "OUT_OF_RANGE";
    CanonicalCodeString["UNIMPLEMENTED"] = "UNIMPLEMENTED";
    CanonicalCodeString["INTERNAL"] = "INTERNAL";
    CanonicalCodeString["UNAVAILABLE"] = "UNAVAILABLE";
    CanonicalCodeString["DATA_LOSS"] = "DATA_LOSS";
    CanonicalCodeString["UNAUTHENTICATED"] = "UNAUTHENTICATED";
})(CanonicalCodeString || (CanonicalCodeString = {}));
class CustomCensusExporter {
    constructor(config) {
        this.transport = serviceManager_1.ServiceManager.get('transport');
        this.config = config;
        this.buffer = new core_1.ExporterBuffer(this, default_config_1.defaultConfig);
    }
    onEndSpan(root) {
        this.buffer.addToBuffer(root);
    }
    onStartSpan(root) { }
    sendTraces(zipkinTraces) {
        return new Promise((resolve, reject) => {
            zipkinTraces.forEach(span => {
                const isRootClient = span.kind === 'CLIENT' && !span.parentId;
                if (isRootClient && this.config.outbound === false)
                    return;
                if (process.env.NODE_ENV === 'test' || (span.duration > constants_1.Constants.MINIMUM_TRACE_DURATION)) {
                    this.transport.send('trace-span', span);
                }
            });
            resolve();
        });
    }
    mountSpanList(rootSpans) {
        const spanList = [];
        for (const root of rootSpans) {
            spanList.push(this.translateSpan(root));
            for (const span of root.spans) {
                spanList.push(this.translateSpan(span));
            }
        }
        return spanList;
    }
    translateSpan(span) {
        const spanTranslated = {
            traceId: span.traceId,
            name: span.name,
            id: span.id,
            parentId: span.parentSpanId,
            kind: this.getSpanKind(span.kind),
            timestamp: span.startTime.getTime() * 1000,
            duration: Math.round(span.duration * 1000),
            debug: false,
            shared: false,
            localEndpoint: { serviceName: this.config.serviceName },
            tags: span.attributes
        };
        if (typeof spanTranslated.tags['result.code'] !== 'string') {
            spanTranslated.tags['result.code'] = CanonicalCodeString[span.status.code];
        }
        if (typeof span.status.message === 'string') {
            spanTranslated.tags['result.message'] = span.status.message;
        }
        return spanTranslated;
    }
    publish(rootSpans) {
        const spanList = this.mountSpanList(rootSpans);
        return this.sendTraces(spanList).catch((err) => {
            return err;
        });
    }
    getSpanKind(kind) {
        switch (kind) {
            case core_1.SpanKind.CLIENT: {
                return 'CLIENT';
            }
            case core_1.SpanKind.SERVER: {
                return 'SERVER';
            }
            default: {
                return 'UNKNOWN';
            }
        }
    }
}
exports.CustomCensusExporter = CustomCensusExporter;
